source-firestore/main.go

```diff
@@ -0,0 +1,411 @@
+					Data:      encDoc.doc,
+				},
+			}); err != nil {
+				return err
+			}
+			if err := enc.Encode(airbyte.Message{
+				Type:  airbyte.MessageTypeState,
+				State: &airbyte.State{Data: []byte("{}"), Merge: true},
+			}); err != nil {
+				return err
+			}
+		} else if len(encDoc.state) != 0 {
+			if err := enc.Encode(airbyte.Message{
+				Type:  airbyte.MessageTypeState,
+				State: &airbyte.State{Data: encDoc.state},
+			}); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func listenForChanges(ctx context.Context, collection *firestore.CollectionRef, docsCh chan encDocument) error {
+	var query = collection.Query
+	var snapshotsIterator = query.Snapshots(ctx)
+
+	log.WithFields(log.Fields{
+		"collection": collection.ID,
+	}).Info("listening for changes on collection")
+
+	for {
+		var snapshot, err = snapshotsIterator.Next()
+
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		for _, change := range snapshot.Changes {
+			if change.Kind == firestore.DocumentAdded || change.Kind == firestore.DocumentModified {
+				log.WithFields(log.Fields{
+					"kind": change.Kind,
+					"id":   change.Doc.Ref.ID,
+				}).Info("received change")
+
+				var doc = change.Doc
+				var data = doc.Data()
+				data[firestore.DocumentID] = doc.Ref.ID
+				docJson, err := json.Marshal(data)
+				if err != nil {
+					return err
+				}
+				log.WithFields(log.Fields{
+					"doc": string(docJson),
+					"id":  change.Doc.Ref.ID,
+				}).Info("received change doc")
+				docsCh <- encDocument{
+					streamName: collection.ID,
+					doc:        docJson,
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func fullScan(ctx context.Context, lastScan time.Time, scanInterval time.Duration, collection *firestore.CollectionRef, docsCh chan encDocument, tail bool) error {
+	var nextScan = time.Until(lastScan.Add(scanInterval))
+	// Wait until next scan
+	time.Sleep(nextScan)
+
+	log.WithFields(log.Fields{
+		"collection": collection.ID,
+	}).Info("full scan on collection")
+
+	var query = collection.Query
+	var docsIterator = query.Documents(ctx)
+
+	var scanTime, _ = time.Now().MarshalText()
+	for {
+		var doc, err = docsIterator.Next()
+
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		var data = doc.Data()
+		data[firestore.DocumentID] = doc.Ref.ID
+		docJson, err := json.Marshal(data)
+		if err != nil {
+			return err
+		}
+		docsCh <- encDocument{
+			streamName: collection.ID,
+			doc:        docJson,
+		}
+	}
+
+	var newState, err = json.Marshal(State{
+		LastScan: string(scanTime),
+	})
+	if err != nil {
+		return fmt.Errorf("marshalling state: %w", err)
+	}
+	docsCh <- encDocument{
+		state: newState,
+	}
+
+	// If we are not tailing, we close the channel after the full scan
+	// to allow the encoding routine to finish so the connector can exit
+	if !tail {
+		close(docsCh)
+		return nil
+	}
+
+	return fullScan(ctx, time.Now(), scanInterval, collection, docsCh, tail)
```

Comment:
Will it matter that the `time.Now()` here is going to be different than the `scanTime` recorded above? I'm guessing probably not, but it seems like it'll make the scan intervals not line up at exactly 12h (or whatever) increments.

Reply:
Good catch!